pennylane/transforms/batch_input.py

```diff
@@ -73,25 +72,39 @@ def circuit(inputs, weights):
     array([0.46230079, 0.73971315, 0.95666004, 0.5355225 , 0.66180948,
             0.44519553, 0.93874261, 0.9483197 , 0.78737918, 0.90866411])>
     """
-    parameters = tape.get_parameters(trainable_only=False)
-
     argnum = tuple(argnum) if isinstance(argnum, (list, tuple)) else (int(argnum),)
 
-    non_trainable, trainable = [], []
-    for idx, param in enumerate(parameters):
-        if idx in argnum:
-            non_trainable.append(param)
-        else:
-            trainable.append(param)
+    trainable_params = tape.get_parameters()
+    all_parameters = sum(
+        (op.parameters if op.parameters != [] else [None] for op in tape.operations), []
+    )
+    argnum_params = [all_parameters[i] for i in argnum]
 
-    if len(np.unique([qml.math.shape(x)[0] for x in non_trainable])) != 1:
+    if any(
+        any(param is train_param for train_param in trainable_params) for param in argnum_params
+    ):
+        raise ValueError(
+            "Batched inputs must be non-trainable. Please make sure that the parameters indexed by "
+            + "'argnum' have 'requires_grad' set to False."
+        )
```

Comment:
Nice! :)
Depending on the interface, the trainability might be deduced from something other that `requires_grad` (e.g. for TensorFlow the `tf.GradientTape` context takes care of this), maybe the error message could be more "flexible" regarding this? Maybe similar to
```suggestion
        raise ValueError(
            "Batched inputs must be non-trainable. Please make sure that the parameters indexed by "
            + "'argnum' are not marked as trainable."
        )
```